source/TWA/Thesis/Chapter3/SearchableTypes.lagda.md

```diff
@@ -0,0 +1,297 @@
+ where
+  px : decidable-predicate 𝓦 X
+  px = p ∘ inr , d ∘ inr
+  x₀ : X
+  x₀ = pr₁ (Sx px)
+  γx : (Σ x ꞉ X , (pr₁ px x holds)) → pr₁ px x₀ holds
+  γx = pr₂ (Sx px)
+  γ : (Σ x ꞉ 𝟘 + X , (p x holds)) → pr₁ px x₀ holds
+  γ (inr x , pix) = γx (x , pix)
+
++-searchable : {X : 𝓤 ̇ } {Y : 𝓥 ̇ }
+             → searchable 𝓦 X
+             → searchable 𝓦 Y
+             → searchable 𝓦 (X + Y)
++-searchable {𝓤} {𝓥} {𝓦} {X} {Y} Sx Sy (p , d)
+ = Cases (d (inl x₀))
+     (λ  px₀ → inl x₀ , λ _ → px₀)
+     (λ ¬px₀ → inr y₀ , γ ¬px₀)
+ where
+  px : decidable-predicate 𝓦 X
+  px = p ∘ inl , d ∘ inl
+  py : decidable-predicate 𝓦 Y
+  py = p ∘ inr , d ∘ inr
+  x₀ : X
+  x₀ = pr₁ (Sx px)
+  γx : Σ x ꞉ X , (pr₁ px x holds) → pr₁ px x₀ holds
+  γx = pr₂ (Sx px)
+  y₀ : Y
+  y₀ = pr₁ (Sy py)
+  γy : Σ y ꞉ Y , (pr₁ py y holds) → pr₁ py y₀ holds
+  γy = pr₂ (Sy py)
+  γ : ¬ (p (inl x₀) holds)
+    → (Σ xy ꞉ (X + Y) , p xy holds)
+    → p (inr y₀) holds
+  γ ¬px₀ (inl x , pix) = 𝟘-elim (¬px₀ (γx (x , pix)))
+  γ ¬px₀ (inr y , piy) = γy (y , piy)
+
+Fin-searchable : (n : ℕ) → Fin n → searchable 𝓦 (Fin n)
+Fin-searchable 1 _ = 𝟘+-searchable 𝟙-searchable
+Fin-searchable (succ (succ n)) _
+ = +-searchable (Fin-searchable (succ n) 𝟎) 𝟙-searchable
+
+equivs-preserve-searchability
+ : {X : 𝓤  ̇ } {Y : 𝓥  ̇}
+ → (f : X → Y)
+ → is-equiv f
+ → searchable 𝓦 X
+ → searchable 𝓦 Y
+equivs-preserve-searchability {𝓤} {𝓥} {𝓦} {X} {Y}
+ f ((g , η) , _) Sx (p , d) = y₀ , γ
+ where
+  px : decidable-predicate 𝓦 X
+  px = p ∘ f , d ∘ f
+  x₀ : X
+  x₀ = pr₁ (Sx px)
+  γx : Σ x ꞉ X , p (f x) holds → p (f x₀) holds
+  γx = pr₂ (Sx px)
+  y₀ : Y
+  y₀ = f x₀
+  γ : Σ y ꞉ Y , p y holds → p y₀ holds
+  γ (y , py) = γx (g y , transport (λ - → p - holds) (η y ⁻¹) py)
+
+≃-searchable
+ : {X : 𝓤  ̇ } {Y : 𝓥 ̇ } → X ≃ Y → searchable 𝓦 X → searchable 𝓦 Y
+≃-searchable (f , e) = equivs-preserve-searchability f e
+             
+finite-searchable : {X : 𝓤 ̇ }
+                  → finite-linear-order X
+                  → X
+                  → searchable 𝓦 X
+finite-searchable (0 , (g , _)) x = 𝟘-elim (g x)
+finite-searchable (succ n , e) x
+ = ≃-searchable (≃-sym e) (Fin-searchable (succ n) 𝟎) 
+
+×-searchable : {X : 𝓤 ̇ } {Y : 𝓥 ̇ }
+             → searchable 𝓦 X
+             → searchable 𝓦 Y
+             → searchable 𝓦 (X × Y)
+×-searchable {𝓤} {𝓥} {𝓦} {X} {Y} Sx Sy (p , d)
+ = xy₀ , γ
+ where
+  py→ : X → decidable-predicate 𝓦 Y
+  py→ x = p ∘ (x ,_) , d ∘ (x ,_)
+  y₀ : X → Y
+  y₀ x = pr₁ (Sy (py→ x))
+  γy : (x : X) → Σ y ꞉ Y , p (x , y) holds → p (x , y₀ x) holds
+  γy x = pr₂ (Sy (py→ x))
+  px : decidable-predicate 𝓦 X
+  px = (λ x → p (x , y₀ x)) , (λ x → d (x , y₀ x))
+  x₀ : X
+  x₀ = pr₁ (Sx px)
+  γx : Σ x ꞉ X , p (x , y₀ x) holds → p (x₀ , y₀ x₀) holds
+  γx = pr₂ (Sx px)
+  xy₀ : X × Y
+  xy₀ = x₀ , y₀ x₀
+  γ : Σ (x , y) ꞉ X × Y , p (x , y) holds → p (x₀ , y₀ x₀) holds
+  γ ((x , y) , pxy) = γx (x , (γy x (y , pxy)))
+```
+
+## Cantor searchability is LPO
+
+```
```

Comment:
A lot of these definitions, constructions and theorems already exist in `TypeTopology`, as you know. I think you should add comments. In general, I think you should also add comments to your own code. Although this is all explained in your thesis, some people may prefer to look at the code directly. A main selling point of `TypeTopology` is that it is a `literate` development. I want to keep it this way.